hydra-node/src/Hydra/Chain/Direct/Handlers.hs

```diff
@@ -67,40 +68,54 @@ import System.IO.Error (userError)
 data LocalChainState m = LocalChainState
   { getLatest :: STM m ChainStateAt
   , pushNew :: ChainStateAt -> STM m ()
-  , rollback :: ChainPoint -> STM m ChainStateAt
+  , rollback :: ChainPoint -> STM m (NonEmpty ChainStateAt)
+  , history :: STM m (NonEmpty ChainStateAt)
   }
 
 -- | Initialize a new local chain state with given 'ChainStateAt' (see also
 -- 'initialChainState').
 newLocalChainState ::
   MonadSTM m =>
-  ChainStateAt ->
+  NonEmpty ChainStateAt ->
   m (LocalChainState m)
-newLocalChainState chainStateAt = do
-  tv <- newTVarIO chainStateAt
+newLocalChainState chainState = do
+  tv <- newTVarIO chainState
   pure
     LocalChainState
-      { getLatest = readTVar tv
+      { getLatest = getLatest tv
       , pushNew = pushNew tv
       , rollback = rollback tv
+      , history = readTVar tv
       }
  where
+  getLatest tv = do
+    currentChainState <- readTVar tv
+    return (head currentChainState)
 
   pushNew tv cs =
-    modifyTVar tv $ \prev ->
-      cs{previous = Just prev}
+    modifyTVar tv $ \prev -> cs :| toList prev
 
   rollback tv point = do
-    latest <- readTVar tv
-    let rolledBack = go point latest
+    history <- readTVar tv
+    let rolledBack = fromMaybe (initialChainState :| []) (go point history)
     writeTVar tv rolledBack
     pure rolledBack
 
-  go rollbackChainPoint = \case
-    cs@ChainStateAt{recordedAt = Just recordPoint}
-      | recordPoint <= rollbackChainPoint -> cs
-    ChainStateAt{previous = Just prev} -> go rollbackChainPoint prev
-    cs -> cs
+  go rollbackChainPoint history =
```

Comment:
We usually name functions `go` when they are recursive. Now this one is not anymore so maybe we could give it more descriptive name?